File: app/frontend/src/lib/campaignCategories.ts
export type CampaignCategoryDef = {
  value: string;
  label: string;
  emoji: string;
  gradient: string;
};

/** Canonical list for create flow, filters, and AI prompts. Backend stores category as a free-form string. */
export const CAMPAIGN_CATEGORIES: CampaignCategoryDef[] = [
  { value: 'environment', label: 'Environment', emoji: '🌱', gradient: 'from-emerald-500 to-green-600' },
  { value: 'health', label: 'Health', emoji: '❤️', gradient: 'from-rose-500 to-pink-600' },
  { value: 'education', label: 'Education', emoji: '📚', gradient: 'from-blue-500 to-indigo-600' },
  { value: 'food', label: 'Food & hunger', emoji: '🍲', gradient: 'from-amber-500 to-orange-600' },
  { value: 'animals', label: 'Animals', emoji: '🐾', gradient: 'from-purple-500 to-violet-600' },
  { value: 'community', label: 'Community', emoji: '🤝', gradient: 'from-teal-500 to-cyan-600' },
  { value: 'arts', label: 'Arts & culture', emoji: '🎭', gradient: 'from-fuchsia-500 to-pink-600' },
  { value: 'sports', label: 'Sports', emoji: '⚽', gradient: 'from-lime-500 to-green-500' },
  { value: 'disaster', label: 'Disaster relief', emoji: '🆘', gradient: 'from-red-600 to-orange-600' },
  { value: 'housing', label: 'Housing', emoji: '🏠', gradient: 'from-sky-500 to-blue-600' },
  { value: 'justice', label: 'Justice & rights', emoji: '⚖️', gradient: 'from-slate-600 to-slate-800' },
  { value: 'innovation', label: 'Tech & innovation', emoji: '💡', gradient: 'from-cyan-500 to-blue-500' },
  { value: 'children', label: 'Children & youth', emoji: '👧', gradient: 'from-pink-400 to-rose-500' },
  { value: 'memorial', label: 'Memorial', emoji: '🕯️', gradient: 'from-stone-500 to-neutral-600' },
  { value: 'faith', label: 'Faith communities', emoji: '🙏', gradient: 'from-violet-400 to-indigo-500' },
  { value: 'veterans', label: 'Veterans', emoji: '🎖️', gradient: 'from-amber-700 to-yellow-600' },
  { value: 'women', label: "Women's causes", emoji: '💜', gradient: 'from-purple-400 to-fuchsia-500' },
];

export const CAMPAIGN_CATEGORY_SLUGS = CAMPAIGN_CATEGORIES.map((c) => c.value).join(', ');

export const CAMPAIGN_CATEGORY_GRADIENTS: Record<string, string> = Object.fromEntries(
  CAMPAIGN_CATEGORIES.map((c) => [c.value, c.gradient])
);

File: app/backend/core/campaign_categories.py
"""
Canonical campaign category slugs for validation (keep in sync with frontend `campaignCategories.ts`).
"""

# fmt: off
CAMPAIGN_CATEGORY_SLUGS: tuple[str, ...] = (
    "environment",
    "health",
    "education",
    "food",
    "animals",
    "community",
    "arts",
    "sports",
    "disaster",
    "housing",
    "justice",
    "innovation",
    "children",
    "memorial",
    "faith",
    "veterans",
    "women",
)
# fmt: on

CAMPAIGN_CATEGORY_SET = frozenset(CAMPAIGN_CATEGORY_SLUGS)

# Common model outputs → slug
_CATEGORY_ALIASES: dict[str, str] = {
    "animal": "animals",
    "wildlife": "animals",
    "hunger": "food",
    "food_security": "food",
    "foodandhunger": "food",
    "food_hunger": "food",
    "tech": "innovation",
    "technology": "innovation",
    "stem": "innovation",
    "startup": "innovation",
    "human_rights": "justice",
    "civil_rights": "justice",
    "legal": "justice",
    "emergency": "disaster",
    "relief": "disaster",
    "culture": "arts",
    "music": "arts",
    "art": "arts",
    "kids": "children",
    "youth": "children",
    "lgbtq": "justice",
    "medical": "health",
    "healthcare": "health",
    "climate": "environment",
    "nature": "environment",
    "local": "community",
    "neighborhood": "community",
    "memorial_fund": "memorial",
    "tribute": "memorial",
}


def normalize_campaign_category(raw: str | None) -> tuple[str, list[str]]:
    """
    Map free-form model output to a whitelist slug.

    Returns:
        (slug, notes) where notes describe coercions for optional UI display.
    """
    notes: list[str] = []
    if not raw or not str(raw).strip():
        return "community", ["category was empty; defaulted to community"]

    s = str(raw).strip().lower().replace("-", "_")
    s = "".join(c if c.isalnum() or c == "_" else "_" for c in s)
    while "__" in s:
        s = s.replace("__", "_")
    s = s.strip("_")

    if s in CAMPAIGN_CATEGORY_SET:
        return s, notes

    if s in _CATEGORY_ALIASES:
        mapped = _CATEGORY_ALIASES[s]
        notes.append(f'category "{raw}" mapped to {mapped}')
        return mapped, notes

    # Longer user strings: only treat as a match if the canonical slug appears as a substring
    # (avoid short false positives like "men" inside "women").
    if len(s) >= 4:
        for slug in CAMPAIGN_CATEGORY_SLUGS:
            if len(slug) >= 4 and slug in s:
                notes.append(f'category "{raw}" matched to {slug}')
                return slug, notes

    notes.append(f'category "{raw}" unknown; defaulted to community')
    return "community", notes
